src/ast.ts

```diff
@@ -1972,9 +1972,24 @@ export function mangleInternalName(declaration: DeclarationStatement, asGlobal:
     return mangleInternalName(<DeclarationStatement>parent, asGlobal) +
            STATIC_DELIMITER + name;
   }
-  return asGlobal
-    ? name
-    : declaration.range.source.internalPath + PATH_DELIMITER + name;
+  return asGlobal ? name : getSourceLevelName(declaration.range.source, name);
+}
+
+export function getSourceLevelName({ internalPath }: Source, simpleName: string): string {
+  return getSourceLevelNameFromInternalPath(internalPath, simpleName);
+}
+
+export function getSourceLevelNameFromInternalPath(internalPath: string, simpleName: string): string {
+  return stripIndex(internalPath) + PATH_DELIMITER + simpleName;
+}
+
+// "foo/index" -> "foo"
+// "foo/bar" -> itself
+export function stripIndex(internalPath: string): string {
```

Comment:
Before, this worked as follows: On `import "some/file"`, the compiler first requested `some/file.ts` from the frontend (here: asc), and if that didn't exist, `some/file/index.ts`. Both had a distinct internal name (either `some/file` or `some/file/index`), and their declarations were not mixed in any way. From a source file, one could both import `some/file` and `some/file/index`, if both would happen to exist. This matches what node.js is doing, and I wonder why a change like stripping `/index` is necessary, effectively removing that functionality?

Reply:
I've updated this branch so we can support `foo.ts` and `foo/index.ts` simultaneously.

There actually was a bug due to being too permissive and trying to resolve imports from both `foo.ts` and `foo/index.ts`:

`foo.ts`: `export const foo1 = 0;`
`foo/index.ts`: `export const foo2 = 0;`
user.ts:
```ts
import {} from "./foo"; //ensure file loaded
import { foo1 } from "./foo/index"; // Shouldn't compile

export function test(): i32 {
  return foo1;
}
```

Shouldn't compile, but did. Fixed in this branch. (Would be nice if we could have negative tests to verify that this isn't allowed, is there an issue open for that?)